ONSdigital/sds
# Sdss 839 dynamic batch size

### Motivation and Context
This change dynamically adjusts the batch size based on the size of unit data (up to a max of 9MB for write and delete), ensuring the system can handle datasets of various sizes without crashing. Previously, batch size was statically limited by count.


### What has changed
- Removed the static limit of 100 units per batch.
- Added logic to calculate the batch size dynamically based on the size of serialized data.
- Committed the batch early if the 9MB limit is reached.
- Reworked deletion process by iterating over documents in batches of 10, dynamically tracking document size to ensure total batch size doesn't exceed 9MB.

### How to test?
- Run the code with a dataset containing various unit sizes.
- Ensure no errors occur during large dataset processing.
- use the link on cloud function performance test to copy cpu, memory configuration and to compare its time.
- you can log it out to confirm batch size is dynamically adjusted and committed when the 9MB limit is reached
- you can force an exception in the perform_batched_dataset_write(...) method after loop to trigger and test the delete batch function. 

### Links
https://jira.ons.gov.uk/browse/SDSS-839
https://confluence.ons.gov.uk/display/SDC/Cloud+Function+Performance+Testing

### Screenshots (if appropriate):
